### PR TITLE
Restore HttpAvailabilityStrategy as default for HttpStreams

### LINKED ISSUE #20401: Roll out AvailabilityStrategy - slowly for GA connectors, as default for others
After having implemented https://github.com/airbytehq/airbyte/pull/19977, in order to better handle 403 errors that cause stream failures to turn into sync failures, we should make sure that GA connectors have the latest CDK installed with the default HttpAvailabilityStrategy available to streams that inherit the HttpStream class. This is an important step that we can do before #17853, which is likely to take more time to implement.

P0: GA connectors that have given us problems in the past: Grab from [this document](https://airbyte.metabaseapp.com/question/1130-permission-failures-by-connector-by-quarter)

P1. Other GA connectors 

P2. Beta/Alpha connectors

Open questions
* if/when do we want to enforce that connectors have this dependency updated? Shouldn't be an issue in the future, but for now we want to get the default behavior in. When moving to GA? When moving to beta? 
   * Idea - maybe must have the default HttpAvailabilityStrategy for beta, must implement own AvailabilityStrategy for GA?
* Most of these will have build fails due to low strictness. How do we handle this, especially if potentially bumping the CDK by a lot? (depends on both dependency and when the source was built last - `docker run pip list`)

### PR DESCRIPTION
**Pull request by erohmensing**
Set HttpAvailabilityStrategy as default for HttpStreams
Reverts https://github.com/airbytehq/airbyte/pull/21488 and updates documentation accordingly
Closes https://github.com/airbytehq/airbyte/issues/20401

Originally was going to be part of https://github.com/airbytehq/airbyte/pull/21888, but want to avoid race conditions where if a connector is not published correctly, it could accidentally get published with the default by someone else before it could be published with the override.

**Don't merge/release until connectors in https://github.com/airbytehq/airbyte/pull/21888 have been published!**

## Pre-Merge
- [ ] All connectors from https://github.com/airbytehq/airbyte/pull/21888 have been published

## Post-Merge follow-up 
- [ ] Publish CDK

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py [hunk 1]
```diff
@@ -113,6 +115,10 @@ def retry_factor(self) -> float:
     def authenticator(self) -> HttpAuthenticator:
         return self._authenticator
 
+    @property
+    def availability_strategy(self) -> Optional[AvailabilityStrategy]:
```
Comment (alafanechere) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py:119]: nit: Would we want a test that asserts the default  `availability_strategy` of an HttpStream is an `HttpAvailabilityStrategy` instance.

**Review by alafanechere — APPROVED** [review]
👍

**Review by evantahler — APPROVED** [review]

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py (hunk 1 above)
Comment (erohmensing) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py:119]: That is handled [here](https://github.com/airbytehq/airbyte/pull/21924/files#diff-0be73143af526e026bfe75f6fc8e166b0431f7fc2ccb847d3410560246d3f24cR61)

```
http_stream = MockHttpStream()
    assert isinstance(http_stream.availability_strategy, HttpAvailabilityStrategy)
```

Where `MockHttpStream()` subclasses `HttpStream` and before this PR, was given the default explicitly:

```
        # TODO (Ella): Remove explicit definition when turning on default
        @property
        def availability_strategy(self) -> Optional["AvailabilityStrategy"]:
            return HttpAvailabilityStrategy()
```
but the explicit definition [has been removed in this PR](https://github.com/airbytehq/airbyte/pull/21924/files#diff-0be73143af526e026bfe75f6fc8e166b0431f7fc2ccb847d3410560246d3f24cL43-L47), so the test now checks that it inherits that default

**Comment by sherifnada** [comment]
asking for a review from one of @brianjlai or @maxi297 to represent ConEx

**Review by sherifnada — APPROVED** [review]

**Review by maxi297 — APPROVED** [review]
LGTM!

File: airbyte-cdk/python/docs/concepts/http-streams.md [hunk 2]
```diff
@@ -87,15 +87,20 @@ be returned as a keyword argument.
...
 
-For HTTP streams, a `HttpAvailabilityStrategy` is defined, which attempts to read the first record of the stream, and excepts
+For HTTP streams, a default `HttpAvailabilityStrategy` is defined, which attempts to read the first record of the stream, and excepts
 a dictionary of known error codes and associated reasons, `reasons_for_unavailable_status_codes`. By default, this list contains only
 `requests.status_codes.FORBIDDEN` (403), with an associated error message that tells the user that they are likely missing permissions associated with that stream.
 
-You can use this `HttpAvailabilityStrategy` in your `HttpStream` by adding the following property to your stream class:
+### Customizing stream availability
+
+You can subclass `HttpAvailabilityStrategy` to override the `reasons_for_unavailable_status_codes` to except more HTTP error codes and inform the user how to resolve errors specific to your connector or stream.
```
[the remark this answers is not in the document]
Comment (erohmensing) [reply airbyte-cdk/python/docs/concepts/http-streams.md:96]: > let's say we refactor the concept of HttpAvailabilityStrategy, we now have to migrate all the custom implementations

This is a good point, and something that I think Alex was trying to get across in a previous review. However, I find a lot of the same problems in the current implementation of having users subclass AbstractSource and HttpStream. I think a lot of the implementations for handling availability will be specific to different sources - being able to provide specific error messages to the users is key to making this an actually useful feature. However, passing that dictionary of statuses and messages as an init parameter does seem like it'd also work and allow for easier refactoring down the line. 

If you're feeling strongly about it, do you want to pair for a short amount of time tomorrow to talk problems and potential solutions?

[the remark this answers is not in the document]
Comment (maxi297) [reply airbyte-cdk/python/docs/concepts/http-streams.md:96]: > I think a lot of the implementations for handling availability will be specific to different sources

For HTTP, I think we could probably get fancy there but I'm wondering if the value it provides is worth the maintenance incurred by subclassing. And yes, other technologies (RPC, SOAP, DB, etc...) will probably need to have different strategies but the less we have, the better it is. 

> If you're feeling strongly about it, do you want to pair for a short amount of time tomorrow to talk problems and potential solutions?

Feel free to book anything that fits in my calendar! I don't have the full context on this and therefore my point might be void once I know more about the situation

[the remark this answers is not in the document]
Comment (erohmensing) [reply airbyte-cdk/python/docs/concepts/http-streams.md:96]: As discussed, we'll move forward with this as planned and take a look at what the implementations look like as GL starts migrating connectors which have custom error handling to the HttpAvailabilityStrategy, with the hopes of getting something with a good default case, and configurable via the low-code parameters, into the low-code CDK 1.0. I created [an issue](https://app.zenhub.com/workspaces/connector-operations-6318ede5eda32e0010ef8566/issues/gh/airbytehq/airbyte/22219) in my team's board to track it, but see it as something we might collaborate on. Feel free to add anything to the issue, especially timelines re: low code release!

**Pull request merged**